substrate/client/network/src/litep2p/shim/request_response/mod.rs

```diff
@@ -0,0 +1,453 @@
+	pub(crate) fn new(
+		protocol_name: ProtocolName,
+		fallback_names: Vec<ProtocolName>,
+		max_request_size: u64,
+		max_response_size: u64,
+		request_timeout: Duration,
+		inbound_queue: Option<async_channel::Sender<IncomingRequest>>,
+	) -> Self {
+		Self {
+			protocol_name,
+			fallback_names,
+			max_request_size,
+			max_response_size,
+			request_timeout,
+			inbound_queue,
+		}
+	}
+}
+
+impl RequestResponseConfigT for RequestResponseConfig {
+	fn protocol_name(&self) -> &ProtocolName {
+		&self.protocol_name
+	}
+}
+
+/// Request-response protocol.
+///
+/// This is slightly different from the `RequestResponsesBehaviour` in that it is protocol-specific,
+/// meaning there is an instance of `RequestResponseProtocol` for each installed request-response
+/// protocol and that instance deals only with the requests and responses of that protocol, nothing
+/// else. It also differs from the other implementation by combining both inbound and outbound
+/// requests under one instance so all request-response-related behavior of any given protocol is
+/// handled through one instance of `RequestResponseProtocol`.
+pub struct RequestResponseProtocol {
+	/// Protocol name.
+	protocol: ProtocolName,
+
+	/// Handle to request-response protocol.
+	handle: RequestResponseHandle,
+
+	/// Inbound queue for sending received requests to protocol implementation in Polkadot SDK.
+	inbound_queue: Option<async_channel::Sender<IncomingRequest>>,
+
+	/// Handle to `Peerstore`.
+	peerstore_handle: Arc<dyn PeerStoreProvider>,
+
+	/// Pending responses.
+	pending_inbound_responses: HashMap<
+		RequestId,
+		(oneshot::Sender<Result<(Vec<u8>, ProtocolName), RequestFailure>>, Instant),
+	>,
+
+	/// Pending outbound responses.
+	pending_outbound_responses: FuturesUnordered<
+		BoxFuture<'static, (litep2p::PeerId, RequestId, Result<OutgoingResponse, ()>, Instant)>,
+	>,
+
+	/// RX channel for receiving info for outbound requests.
+	request_rx: TracingUnboundedReceiver<OutboundRequest>,
+
+	/// Metrics, if enabled.
+	metrics: RequestResponseMetrics,
+}
+
+impl RequestResponseProtocol {
+	/// Create new [`RequestResponseProtocol`].
+	pub fn new(
+		protocol: ProtocolName,
+		handle: RequestResponseHandle,
+		peerstore_handle: Arc<dyn PeerStoreProvider>,
+		inbound_queue: Option<async_channel::Sender<IncomingRequest>>,
+		metrics: Option<Metrics>,
+	) -> (Self, TracingUnboundedSender<OutboundRequest>) {
+		let (request_tx, request_rx) = tracing_unbounded("outbound-requests", 10_000);
+
+		(
+			Self {
+				handle,
+				request_rx,
+				inbound_queue,
+				peerstore_handle,
+				protocol: protocol.clone(),
+				pending_inbound_responses: HashMap::new(),
+				pending_outbound_responses: FuturesUnordered::new(),
+				metrics: RequestResponseMetrics::new(metrics, protocol),
+			},
+			request_tx,
+		)
+	}
+
+	/// Send `request` to `peer`.
+	async fn on_send_request(
+		&mut self,
+		peer: PeerId,
+		request: Vec<u8>,
+		// TODO: discuss the design of fallback requests
```

Comment:
My main issue is that there is no strict definition (or I can't find it) whether the old, binary-incompatible `/request/1` is a fallback protocol of the new `/request/2` protocol, i.e., is the old protocol put into `fallback_names` in the request-response protocol configuration. The test you wrote in the PR doesn't use the `fallback_names` field which would indicate that the new protocol is considered a completely new protocol and should have its own `RequestResponseProtocol` in the litep2p backend. If this is the case, I have to make a hack in the backend or redesign the request-response protocol support for the new backend which isn't the end of the world.

If the old protocol is considered a "true" fallback in `multistream-select` sense, where it tries to negotiate the protocol using one handshake, I'd need to add a new function to litep2p which would be considerably less work. This could be possible to implement for the libp2p backend as well, since the negotiated protocol [is available](https://github.com/libp2p/rust-libp2p/blob/ddb554440267c5d441e47827c2aa83e11b171e1d/protocols/request-response/src/codec.rs#L61) when the request is sent. I did not think of this when the original PR was under review.

IOW, I intentionally postponed implementing the fallback support (which I'll implement before this PR done) until I understood how the fallback requests should be implemented.

Reply:
It wouldn't even necessarily have to be a new codec. We could tweak the current codec implementation to take `(Vec<u8>, Option<(Vec<u8>, ProtocolName>)` as parameter and once it's called, check which protocol was negotiated and see if the fallback request is `Some`.

Which direction should we take?

Reply:
I'll play around with this a bit. It would be great if we could just use multi-stream select as it was intended to

Reply:
I was playing around with [some code](https://github.com/paritytech/polkadot-sdk/commit/cbd7dda92bc079bdccd3e8f809fb25a26cb663d8) last night that would allow using `multistream-select` to do the negotiation but the APIs we're provided with are quite unforgiving and it's not possible to, e.g., return a custom error from `Codec` so if you misuse the API and forget to provide a fallback request for a peer who doesn't support the new protocol, it will kill the connection. That code is not fully tested and can obviously be refactored a bit but I think it's really ugly compared to current fallback support.

There's one problem with `multistream-select` though and that's mixing two types of fallbacks. If we have `/<genesis hash>/request/2` and `/polkadot/request/2` which are binary-incompatible with `/<genesis hash>/request/1` and `/polkadot/request/1`, we have to have a mapping between protocol names and their requests because `/<genesis hash>/request/1` and `/polkadot/request/1` are fallbacks for `/<genesis hash/request/2` but `/polkadot/request/1` is also a fallback for `/<genesis hash>/polkadot/1` which complicates things. As far as I see it, we'd have to have a `HashMap<ProtocolName, Vec<u8>` for each supported protocol.

If we want to keep using fallback names (`/polkadot/` prefix) for our protocols then I'd be in favor of closing this discussion and going with your approach because the benefits of supporting two different types of fallbacks with `multistream-select` are not IMO worth the implementation and maintenance costs of shaving one handshake from the fallback support but that's of course debatable.

Reply:
looking at your code, I think that could work indeed

> There's one problem with multistream-select though and that's mixing two types of fallbacks. If we have /<genesis hash>/request/2 and /polkadot/request/2 which are binary-incompatible with /<genesis hash>/request/1 and /polkadot/request/1, we have to have a mapping between protocol names and their requests because /<genesis hash>/request/1 and /polkadot/request/1 are fallbacks for /<genesis hash/request/2 but /polkadot/request/1 is also a fallback for /<genesis hash>/polkadot/1 which complicates things. As far as I see it, we'd have to have a HashMap<ProtocolName, Vec<u8> for each supported protocol.

yeah, from what I can tell, this is what you coded in the linked commit, right?

but indeed, in rust-libp2p, looks like it would be quite challenging to expose a fool-proof API.

> If we want to keep using fallback names (/polkadot/ prefix) for our protocols then I'd be in favor of closing this discussion and going with your approach because the benefits of supporting two different types of fallbacks with multistream-select are not IMO worth the implementation and maintenance costs of shaving one handshake from the fallback support but that's of course debatable.

For newly added protocols, we don't add support for the legacy (non-genesishash) name. Because if new nodes want to add support for a new protocol, they might as well not use the legacy name. I don't know for how long we'd need to keep the legacy naming for the existing protocols. I guess we still need to accomodate for non-upgraded nodes.

Another option would be to use a hybrid approach, keep the fallback request as it is for libp2p but design it in a better way for litep2p. Do you have an idea of how that'd look like?

Reply:
I implemented initial support for the fallback support, [here](https://github.com/altonen/litep2p/commit/8a53fff3504be270e7ee63f85da45c915071de4a) and [here](https://github.com/altonen/litep2p/commit/4c4f9445af73205e13faaa95b108e2ef46daef93) are relevant litep2p commits. The tests have a [`FIXME`](https://github.com/paritytech/polkadot-sdk/blob/0396c06a76652584130b777d0091da1ae52b7844/substrate/client/network/src/litep2p/shim/request_response/tests.rs#L615-L620) because `IncomingRequest` doesn't inform the relevant request handler which protocol was negotiated during handshake since it assumes it will always receive a request over the main protocol or one of the negotiated fallbacks, all of which are binary-compatible with each other.

I think this high-level API is incompatible with the idea of using `multistream-select` for fallback requests since the current assumption is that it doesn't need the protocol name to decode the request, which in turn is incompatible with the idea of binary-incompatible requests. We can also tweak the Polkadot-side implementation for litep2p and give it a copy of all inbound request channels and when a request is received, it would send the request over the appropriate channel.

Other options are reverting the litep2p changes and going with the libp2p approach of having separate protocols for each incompatible protocol or modifying the libp2p protocol to use `multistream-select` for fallback support.

Reply:
can't we add the negotiated protocol to the `IncomingRequest`?

Then add an extra generic type parameter to `IncomingRequestReceiver` and `IncomingRequest`, called `FallbackReq` which would by default be identical to `Req`. Then, in `IncomingRequest::try_from_raw()` we check the reported protocol name against the protocol names of the generics before decoding

would smth like this be possible?

The slight drawback would be that for libp2p we'd have multiple request receivers, which are versioned. and for litep2p we'd only have one for both versions. but maybe we could abstract it a bit and make this a bit nicer to work with. Having an IncomingVersionedRequestReceiver which would know how to demultiplex the combined messages of litep2p into separate streams (or the other way around)

Reply:
Sorry for delayed reply, I will have to look into this some more because what you're proposing sounds plausible. One problem with using different approachces (libp2p no `multistream-select`, litep2p and `multistream-select`) is that we're moving complexity to the request handlers/high-level code. Another issue is with initialization because litep2p expects the fallback protocols to be put into `fallback_names` whereas libp2p expects them to be separate protocols which means the initialization is also more complicated with a versioned receiver which demultiplexes the request to correct handler like you mentioned.

I would be inclined to going with a simplified approach where neither library uses `multistream-select` and if we jettison libp2p at some point, we can simplify the APIs and start using `multistream-select`. But I'll experiment with your idea first

Reply:
yeah, I agree. Let's see how much complexity it adds.
If it's too much, we can stick to the simplified approach you mentioned. Your call

Reply:
I played around with your proposal but decided to go with the simpler approach for now because while yours was better in terms of utilizing `multistream-select` correctly, it required quite a bit of changes to existing code, both in Substrate and in Polkadot. Going with the simpler approach won't make things worse since we're already using `V1` and the benefit is that none of the existing code has to be touched. If the fallback support with `multistream-select` is deemed superior and worth the refactoring costs, that code will be available in this PR's history. New implementation here: https://github.com/paritytech/polkadot-sdk/pull/2944/commits/439e65e808fb1fdd92581ff43afd8d736ee4d34e

Reply:
sounds good to me